File: ArryLoad/SignalTypes/DutyCycleSignal.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace ArryLoad.SignalTypes
{
    class DutyCycleSignal : Signal
    {
        public double DutyFactor { get; protected set; }
        public DutyCycleSignal(double A, double f, double fi, double dutyFactor)
            : base(A, f, fi)
        {
            this.DutyFactor = dutyFactor;
        }

        public override double GetValue(int x, int N)
        {

            double deg = Fi * Math.PI / 180;
            int coef = (Math.Sin(2 * Math.PI * F * x/N + deg) + 1 >= 2 - 2 * DutyFactor) ? 1 : 0;
            return A * coef;
        }

        public override double GetValue(int x)
        {
            int N = SampleRate;
            double deg = Fi * Math.PI / 180;
            int coef = (Math.Sin(2 * Math.PI * F * x / N + deg) + 1 >= 2 - 2 * DutyFactor) ? 1 : 0;
            return A * coef; ;
        }
    }
}

File: ArryLoad_VB/ArryLoad.SignalTypes/DutyCycleSignal.vb
Imports System

Namespace ArryLoad.SignalTypes
	Friend Class DutyCycleSignal
		Inherits Signal

		Public Property DutyFactor() As Double

		Public Sub New(A As Double, f As Double, fi As Double, dutyFactor As Double)
			MyBase.[New](A, f, fi)
			AddressOf Me.DutyFactor = dutyFactor
		End Sub

		Public Overrides Function GetValue(x As Integer, N As Integer) As Double
			Dim num As Double = AddressOf MyBase.Fi * 3.1415926535897931 / 180.0
			Dim num2 As Integer = If((Math.Sin(6.2831853071795862 * AddressOf MyBase.F * CDec(x) / CDec(N) + num) + 1.0 >= 2.0 - 2.0 * AddressOf Me.DutyFactor), 1, 0)
			Return AddressOf MyBase.A * CDec(num2)
		End Function

		Public Overrides Function GetValue(x As Integer) As Double
			Dim sampleRate As Integer = AddressOf MyBase.SampleRate
			Dim num As Double = AddressOf MyBase.Fi * 3.1415926535897931 / 180.0
			Dim num2 As Integer = If((Math.Sin(6.2831853071795862 * AddressOf MyBase.F * CDec(x) / CDec(sampleRate) + num) + 1.0 >= 2.0 - 2.0 * AddressOf Me.DutyFactor), 1, 0)
			Return AddressOf MyBase.A * CDec(num2)
		End Function
	End Class
End Namespace
